python/openchange/migration/directory.py

```diff
@@ -0,0 +1,186 @@
+searchFlags: 1
+lDAPDisplayName: msExchRecipientDisplayType
+name: ms-Exch-Recipient-Display-Type
+schemaIDGUID:: sKuTuH52IE+RXzhXu8lq/g==
+attributeSecurityGUID:: iYopH5jeuEe1zVcq1T0mfg==
+isMemberOfPartialAttributeSet: TRUE
+objectCategory: CN=Attribute-Schema,%(schema_dn)s
+
+#
+# ms-Exch-Recipient-Type-Details
+#
+dn: CN=ms-Exch-Recipient-Type-Details,%(schema_dn)s
+objectClass: top
+objectClass: attributeSchema
+cn: ms-Exch-Recipient-Type-Details
+distinguishedName: CN=ms-Exch-Recipient-Type-Details,%(schema_dn)s
+instanceType: 4
+whenCreated: 20140829141013.0Z
+whenChanged: 20140829141013.0Z
+uSNCreated: 43693
+attributeID: 1.2.840.113556.1.4.7000.102.50855
+attributeSyntax: 2.5.5.16
+isSingleValued: TRUE
+uSNChanged: 43693
+showInAdvancedViewOnly: TRUE
+adminDisplayName: ms-Exch-Recipient-Type-Details
+adminDescription: ms-Exch-Recipient-Type-Details
+oMSyntax: 65
+searchFlags: 1
+lDAPDisplayName: msExchRecipientTypeDetails
+name: ms-Exch-Recipient-Type-Details
+schemaIDGUID:: +KGbBgpUqUK/JqfdNUdTRg==
+attributeSecurityGUID:: iYopH5jeuEe1zVcq1T0mfg==
+isMemberOfPartialAttributeSet: TRUE
+objectCategory: CN=Attribute-Schema,%(schema_dn)s
+"""
+
+        add_schema_attr_ldif = add_schema_attr_ldif % {'schema_dn': schema_dn}
+        try:
+            db.add_ldif(add_schema_attr_ldif, ['relax:0'])
+        except Exception, ex:
+            print "Error adding msExchRecipientTypeDetails and msExchRecipientDisplayType atributes to schema: %s" % str(ex)
+
+        force_schemas_update(db, setup_path)
+
+        upgrade_schema_ldif = """
+dn: CN=Mail-Recipient,%(schema_dn)s
+changetype: modify
+add: mayContain
+mayContain: msExchRecipientDisplayType
+mayContain: msExchRecipientTypeDetails
+"""
+        upgrade_schema_ldif = upgrade_schema_ldif % {'schema_dn': schema_dn}
+        try:
+            db.modify_ldif(upgrade_schema_ldif)
+        except Exception, ex:
+            print "Error adding msExchRecipientTypeDetails and msExchRecipientDisplayType to Mail-Recipient schema: %s" % str(ex)
+
+        force_schemas_update(db, setup_path)
+
+        # Add missing attributes for users
+        user_upgrade_ldif = ""
+        user_upgrade_ldif_template = """
+dn: %(user_dn)s
+add: msExchRecipientTypeDetails
+msExchRecipientTypeDetails: 6
+add: msExchRecipientDisplayType
+msExchRecipientDisplayType: 0
+"""
+        base_dn = "CN=Users,%s" % names.domaindn
+        ldb_filter = "(objectClass=user)"
+        res = db.search(base=base_dn, scope=ldb.SCOPE_SUBTREE, expression=ldb_filter)
+        for element in res:
+            dn = element.dn.get_linearized()
+            ldif = user_upgrade_ldif_template % {'user_dn': dn}
+            user_upgrade_ldif += ldif + "\n"
+            try:
+                db.modify_ldif(ldif)
+            except Exception, ex:
+                print "Error migrating user %s: %s. Skipping user" % (dn, str(ex))
+
+    @classmethod
+    def unapply(cls, cur):
+        raise Exception("Cannot revert directory migrations")
```

Comment:
As a developer, I think we should provide the sysadmin the ability to perform this operation with its consequences. In order to perform the backwards migration, the sysadmin should really know what he's doing because it is a difficult path.
In addition to this, it will be as uniform as DB migrations are and they allow us to go backwards for a reason. And finally, it is the only way I see to test your code is resilient enough. Go forwards and backwards twice.

I don't see why you need to remove the full element but only the attributes added by this schema... Could you explain me the first sentence?